storage/fanout.go

```diff
@@ -231,27 +232,57 @@ func NewMergeQuerier(primaryQuerier Querier, queriers []Querier) Querier {
 func (q *mergeQuerier) Select(params *SelectParams, matchers ...*labels.Matcher) (SeriesSet, Warnings, error) {
 	seriesSets := make([]SeriesSet, 0, len(q.queriers))
 	var warnings Warnings
-	for _, querier := range q.queriers {
-		set, wrn, err := querier.Select(params, matchers...)
-		q.setQuerierMap[set] = querier
-		if wrn != nil {
-			warnings = append(warnings, wrn...)
-		}
-		if err != nil {
-			q.failedQueriers[querier] = struct{}{}
-			// If the error source isn't the primary querier, return the error as a warning and continue.
-			if querier != q.primaryQuerier {
-				warnings = append(warnings, err)
-				continue
-			} else {
+	if len(q.queriers) > 1 {
+		wg := sync.WaitGroup{}
+		wg.Add(len(q.queriers))
+		var priErr error = nil
+		var lock sync.Mutex
+		for _, qr := range q.queriers {
+			go func(querier Querier) {
+				defer wg.Done()
+				set, wrn, selectError := querier.Select(params, matchers...)
```

Comment:
I was thinking more that you take this result and send it down a channel.

Reply:
done,move to https://github.com/prometheus/prometheus/pull/6768